File: perl/bin/buildDJobPropFiles4Craig.pl
#!/usr/bin/perl -w
($] >= 5.004) || die "version is $] -- need perl 5.004 or greater";
# -*- perl -*-

use lib "$ENV{CRAIG_HOME}/perl/lib";
use Organism;
use Contig;
use Getopt::Long;
use strict;

my $prefix_propdir;
my $species = "tgondii";
my $dataset_id = "oocyst_booth";
my $sample_id = "day0";
my $rsq_inputdir = "/gpfs/fs121/h/abernal/Applications/Contrib/rum/data/me49-8.0.oocyst_boothday0";
my $prefix_craig_output = "/gpfs/fs121/h/abernal/Data/me49-8.0/me49-8.0.oocyst_boothday0";
my $fasta_file = "/gpfs/fs121/h/abernal/Data/me49-8.0/TgondiiME49_ToxoDB-8.0.gff";
my $annot_file = "/gpfs/fs121/h/abernal/Data/me49-8.0/ToxoDB-8.0_TgondiiME49_Genome.fasta";
my $rsq_orientation = "N";
my $transcript_tag = "exon";
my $num_djobnodes = 60;
my $rsq_type = "rum";
my $closest_species = "";
my $usage = "usage : $0 [Options]\n\t--rsq-type RSQ_TYPE\tEither gsnap or rum [rum]\n\t--rsq-orientation [N|F|R]\n\t--species\tSPECIES\n\t--prefix-propdir\tPREFIX_PROPDIR\n\t--dataset-id\tDATASET_ID\n\t--sample-id\tSAMPLE_ID\n\t--rsq-inputdir\tRSQ_INPUTDIR\n\t--prefix-craig-output\tPREFIX_CRAIG_OUTPUT\n\t--fasta-file\tFASTA_FILE\n\t--annot-file\tANNOT_FILE\n\t--closest-species\tCLOSEST_SPECIES\n";

&GetOptions("--prefix-propdir=s" => \$prefix_propdir,
	    "--rsq-type=s" => \$rsq_type,
	    "--species=s" => \$species,
	    "--dataset-id=s" => \$dataset_id,
	    "--sample-id=s" => \$sample_id,
	    "--rsq-inputdir=s" => \$rsq_inputdir,
	    "--rsq-orientation=s" => \$rsq_orientation,
	    "--prefix-craig-output=s" => \$prefix_craig_output,
	    "--fasta-file=s" => \$fasta_file,
	    "--annot-file=s" => \$annot_file,
	    "--transcript-tag=s" => \$transcript_tag,
	    "--num-djobnodes=i" => \$num_djobnodes,
	    "--closest-species=s" => \$closest_species,
    );

$prefix_propdir ne "" || die "PREFIX_DIR must be given\n".$usage;

my $prop_dir =  $prefix_propdir."_input";
my $djob_dir = $prefix_propdir."_DJob";

`rm -rf $prop_dir` if -d $prop_dir;
`rm -rf $djob_dir` if -d $djob_dir;
`mkdir $prop_dir`;
`mkdir $djob_dir`;

my $task_props = "RSqType=$rsq_type\nRSqOrientation=$rsq_orientation\nRSqWrapperIn=cat\nInputAnnotFormat=gff3\nInputContigFormat=fasta\ntranscriptTag=$transcript_tag\nCDSTag=CDS\ngcClasses=100\ngeneModelType=ngscraig\nmodelUTRs=yes\ngenUTROnlyModel=yes\nnumDJobNodes=$num_djobnodes\nDJobNodeClass=SgeNode\nminPercJunctionAlignments=75\nspecies=$species".($closest_species eq "" ? "" : "\nclosestSpecies=$closest_species")."\nRSqDataSetID=$dataset_id\nRSqSampleID=$sample_id\nRSqInputDir=$rsq_inputdir\nprefixOutputDir=$prefix_craig_output\nInputAnnotFile=$annot_file\nInputContigFile=$fasta_file\n";

my $controller_props = "masterdir=$prop_dir/master\ninputdir=$prop_dir\nnodedir=$djob_dir\nslotspernode=1\nsubtasksize=1\ntaskclass=DJob::DistribJobTasks::CRAIGTask\nnodeclass=DJob::DistribJob::SgeNode\n";

open(TASK_PROP, ">$prop_dir/task.prop") || die "Can't open $prop_dir/task.prop";
print TASK_PROP $task_props;
close(TASK_PROP);
open(CONTROLLER_PROP, ">$prop_dir/controller.prop") || die "Can't open $prop_dir/controller.prop";
print CONTROLLER_PROP $controller_props;
close(CONTROLLER_PROP);

File: python/bin/buildDJobPropFiles4Craig.py
#!/usr/bin/env python

import struct, argparse, string, sys, numpy, os, warnings, subprocess, os.path, shutil, logging, re
sys.path.append(os.environ['CRAIG_HOME']+'/python/lib/')

import bioutils, ioutils
parser = argparse.ArgumentParser(description='Script for building task.prop, conttroller.prop and all the associated directories for executing the CRAIG v.1.1 pipeline, a discriminative gene prediction tool for eukarya.\nWritten by Axel E. Bernal (abernal@seas.upenn.edu).')

parser.add_argument("SPECIES", type=str, help='Species name, given by the common suffix of files present in the $CRAIG_HOME/models directory. These files specify the resources, filters and features used for constructing gene models and will be generated. Possible values right now are "tgondii" "cvelia" and "vitrella"', action='store')
parser.add_argument("ANNOT_FILE", type=str, help='Name of file with gene annotations', action='store')
parser.add_argument("FASTA_FILE", type=str, help='Name of file with fasta sequences or contigs', action='store')
parser.add_argument("RSQ_INPUTDIR", type=str, help='Mapped RNA-Seq input directory, i.e. where the RNA-Seq mapped data is located"', action='store')
parser.add_argument("PREFIX_CRAIG_OUTPUT", type=str, help='Prefix for CRAIG output files. Output will be found in PREFIX_CRAIG_OUTPUT.preproc and PREFIX_CRAIG_OUTPUT.model. Temporary files for DJob execution will be found in PREFIX_CRAIG_OUTPUT.test. The latter should be erased manually after results are obtained', action='store')
parser.add_argument("PREFIX_PROPDIR", type=str, help='Distribjob input directory, i.e. where task.prop and controller.prop are going to be located', action='store')
parser.add_argument("--transcript-tag", type=str, help='The description tag as it appears in ANNOT_FILE for transcript regions(including utrs). If equal to null, it is assumed that the annotations only contain CDS regions', default='null')
parser.add_argument("--rsq-type", type=str, help='Type of RNA-Seq alignment. Either RUM or GSNAP are currently supported', default='gsnap')
parser.add_argument("--cds-tag", type=str, help='The description tag as it appears in ANNOT_FILE for coding regions.', default='CDS')
parser.add_argument("--djob-num-nodes", type=int, help='A number greater or equal than 2 would allow the use of distribjob for parallelizing expensive tasks', default=40)
parser.add_argument("--dataset-id", type=str, help='RNA-Seq dataset id. Typically contains or makes reference to the organism and author names. For example "oocyst_booth"', default='oocyst_booth')
parser.add_argument("--sample-id", type=str, help='RNA-Seq sample id. Typically contains or makes reference to the condition, tissue, stage of development in which the reads were generated. For example "day3"', default='day0')
parser.add_argument("--rsq-orientation", type=str, help='Has a value of "N","F" or "R" if the RNA-Seq reads have orientation or not. For paired reads these values must be given in pairs, i.e. "NN", "FR" and so on', default="N")
line  = "";

try:
    args = parser.parse_args(sys.argv[1:])
    prop_dir =  args.PREFIX_PROPDIR+"_input";
    djob_dir = args.PREFIX_PROPDIR+"_DJob";
    log_file = open(args.PREFIX_PROPDIR+".error.log", "w")

    if os.path.exists(prop_dir) :
        shell_cmd = "rm -rf "+prop_dir
        ioutils.log_and_exec(shell_cmd, os.environ, log_file)

    if os.path.exists(djob_dir) :
        shell_cmd = "rm -rf "+djob_dir
        ioutils.log_and_exec(shell_cmd, os.environ, log_file)

    shell_cmd = "mkdir "+prop_dir
    ioutils.log_and_exec(shell_cmd, os.environ, log_file)
    shell_cmd = "mkdir "+djob_dir
    ioutils.log_and_exec(shell_cmd, os.environ, log_file)

    task_props = "RSqType="+args.rsq_type+"\nRSqOrientation="+args.rsq_orientation+"\nRSqWrapperIn=cat\nInputAnnotFormat=gff3\nInputContigFormat=fasta\ntranscriptTag="+args.transcript_tag+"\nCDSTag="+args.cds_tag+"\ngcClasses=100\ngeneModelType=ngscraig\nmodelUTRs=yes\ngenUTROnlyModel=yes\nnumDJobNodes="+str(args.djob_num_nodes)+"\nDJobNodeClass=SgeNode\nminPercJunctionAlignments=75\nspecies="+args.SPECIES+"\nRSqDataSetID="+args.dataset_id+"\nRSqSampleID="+args.sample_id+"\nRSqInputDir="+args.RSQ_INPUTDIR+"\nprefixOutputDir="+args.PREFIX_CRAIG_OUTPUT+"\nInputAnnotFile="+args.ANNOT_FILE+"\nInputContigFile=args.FASTA_FILE\n"

    controller_props = "masterdir="+prop_dir+"/master\ninputdir="+prop_dir+"\nnodedir="+djob_dir+"\nslotspernode=1\nsubtasksize=1\ntaskclass=DJob::DistribJobTasks::CRAIGTask\nnodeclass=DJob::DistribJob::SgeNode\n"
        
    taskprop_fd = open(prop_dir+"/task.prop", "w")
    taskprop_fd.write(task_props)
    taskprop_fd.close()
    controllerprop_fd = open(prop_dir+"/controller.prop", "w")
    controllerprop_fd.write(controller_props)
    controllerprop_fd.close()
except ioutils.CraigUndefined, msg:
    print msg
